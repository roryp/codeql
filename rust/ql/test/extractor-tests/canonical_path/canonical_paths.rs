mod a {
    #[derive(Eq, PartialEq)]
    pub struct Struct;

    pub trait Trait {
        fn f(&self);
    }

    impl Trait for Struct {
        fn f(&self) {}
    }

    impl Struct {
        fn g(&self) {}
    }

    trait TraitWithBlanketImpl {
        fn h(&self);
    }

    impl<T: Eq> TraitWithBlanketImpl for T {
        fn h(&self) {}
    }

    fn free() {}

    fn usage() {
        let s = Struct {};
        s.f();
        s.g();
        s.h();
        free();
    }
}

mod without {
    use super::a::Trait;

    fn canonicals() {
        struct OtherStruct;

        trait OtherTrait {
            fn g(&self);
        }

        impl OtherTrait for OtherStruct {
            fn g(&self) {}
        }

        impl OtherTrait for crate::canonical_paths::a::Struct {
            fn g(&self) {}
        }

        impl crate::canonical_paths::a::Trait for OtherStruct {
            fn f(&self) {}
        }

        fn nested() {
            struct OtherStruct;
        }

        fn usage() {
            let s = OtherStruct {};
            s.f();
            s.g();
            nested();
        }
    }

    fn other() {
        struct OtherStruct;
    }
}
