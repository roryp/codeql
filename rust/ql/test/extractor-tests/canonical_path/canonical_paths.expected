canonicalPaths
| canonical_paths.rs:1:1:34:1 | Module | repo::test | crate::canonical_paths::a |
| canonical_paths.rs:2:5:3:22 | Struct | repo::test | crate::canonical_paths::a::Struct |
| canonical_paths.rs:5:5:7:5 | Trait | repo::test | crate::canonical_paths::a::Trait |
| canonical_paths.rs:6:9:6:20 | f | repo::test | crate::canonical_paths::a::Trait::f |
| canonical_paths.rs:9:5:11:5 | Impl | None | None |
| canonical_paths.rs:10:9:10:22 | f | repo::test | <crate::canonical_paths::a::Struct as crate::canonical_paths::a::Trait>::f |
| canonical_paths.rs:13:5:15:5 | Impl | None | None |
| canonical_paths.rs:14:9:14:22 | g | repo::test | <crate::canonical_paths::a::Struct>::g |
| canonical_paths.rs:17:5:19:5 | Trait | repo::test | crate::canonical_paths::a::TraitWithBlanketImpl |
| canonical_paths.rs:18:9:18:20 | h | repo::test | crate::canonical_paths::a::TraitWithBlanketImpl::h |
| canonical_paths.rs:21:5:23:5 | Impl | None | None |
| canonical_paths.rs:22:9:22:22 | h | repo::test | <_ as crate::canonical_paths::a::TraitWithBlanketImpl>::h |
| canonical_paths.rs:25:5:25:16 | free | repo::test | crate::canonical_paths::a::free |
| canonical_paths.rs:27:5:33:5 | usage | repo::test | crate::canonical_paths::a::usage |
| canonical_paths.rs:36:1:73:1 | Module | repo::test | crate::canonical_paths::without |
| canonical_paths.rs:37:5:37:24 | Use | None | None |
| canonical_paths.rs:39:5:68:5 | canonicals | repo::test | crate::canonical_paths::without::canonicals |
| canonical_paths.rs:40:9:40:27 | Struct | repo::test | {34}::OtherStruct |
| canonical_paths.rs:42:9:44:9 | Trait | repo::test | {34}::OtherTrait |
| canonical_paths.rs:43:13:43:24 | g | repo::test | {34}::OtherTrait::g |
| canonical_paths.rs:46:9:48:9 | Impl | None | None |
| canonical_paths.rs:47:13:47:26 | g | repo::test | <{34}::OtherStruct as {34}::OtherTrait>::g |
| canonical_paths.rs:50:9:52:9 | Impl | None | None |
| canonical_paths.rs:51:13:51:26 | g | repo::test | <crate::canonical_paths::a::Struct as {34}::OtherTrait>::g |
| canonical_paths.rs:54:9:56:9 | Impl | None | None |
| canonical_paths.rs:55:13:55:26 | f | repo::test | <{34}::OtherStruct as crate::canonical_paths::a::Trait>::f |
| canonical_paths.rs:58:9:60:9 | nested | repo::test | {34}::nested |
| canonical_paths.rs:59:13:59:31 | Struct | repo::test | {35}::OtherStruct |
| canonical_paths.rs:62:9:67:9 | usage | repo::test | {34}::usage |
| canonical_paths.rs:70:5:72:5 | other | repo::test | crate::canonical_paths::without::other |
| canonical_paths.rs:71:9:71:27 | Struct | repo::test | {36}::OtherStruct |
| lib.rs:1:1:1:20 | Module | repo::test | crate::canonical_paths |
resolvedPaths
| canonical_paths.rs:2:7:2:12 | derive | None | None |
| canonical_paths.rs:9:10:9:14 | Trait | repo::test | crate::canonical_paths::a::Trait |
| canonical_paths.rs:9:20:9:25 | Struct | repo::test | crate::canonical_paths::a::Struct |
| canonical_paths.rs:13:10:13:15 | Struct | repo::test | crate::canonical_paths::a::Struct |
| canonical_paths.rs:21:13:21:14 | Eq | lang:core | crate::cmp::Eq |
| canonical_paths.rs:21:17:21:36 | TraitWithBlanketImpl | repo::test | crate::canonical_paths::a::TraitWithBlanketImpl |
| canonical_paths.rs:21:42:21:42 | T | None | None |
| canonical_paths.rs:28:17:28:22 | Struct | repo::test | crate::canonical_paths::a::Struct |
| canonical_paths.rs:29:9:29:9 | s | None | None |
| canonical_paths.rs:29:9:29:13 | ... .f(...) | repo::test | <crate::canonical_paths::a::Struct as crate::canonical_paths::a::Trait>::f |
| canonical_paths.rs:30:9:30:9 | s | None | None |
| canonical_paths.rs:30:9:30:13 | ... .g(...) | repo::test | <crate::canonical_paths::a::Struct>::g |
| canonical_paths.rs:31:9:31:9 | s | None | None |
| canonical_paths.rs:31:9:31:13 | ... .h(...) | repo::test | <_ as crate::canonical_paths::a::TraitWithBlanketImpl>::h |
| canonical_paths.rs:32:9:32:12 | free | repo::test | crate::canonical_paths::a::free |
| canonical_paths.rs:37:9:37:13 | super | repo::test | crate::canonical_paths |
| canonical_paths.rs:37:9:37:16 | super::a | repo::test | crate::canonical_paths::a |
| canonical_paths.rs:37:9:37:23 | super::a::Trait | repo::test | crate::canonical_paths::a::Trait |
| canonical_paths.rs:46:14:46:23 | OtherTrait | repo::test | {34}::OtherTrait |
| canonical_paths.rs:46:29:46:39 | OtherStruct | repo::test | {34}::OtherStruct |
| canonical_paths.rs:50:14:50:23 | OtherTrait | repo::test | {34}::OtherTrait |
| canonical_paths.rs:50:29:50:33 | crate | None | None |
| canonical_paths.rs:50:29:50:50 | crate::canonical_paths | repo::test | crate::canonical_paths |
| canonical_paths.rs:50:29:50:53 | crate::canonical_paths::a | repo::test | crate::canonical_paths::a |
| canonical_paths.rs:50:29:50:61 | crate::canonical_paths::a::Struct | repo::test | crate::canonical_paths::a::Struct |
| canonical_paths.rs:54:14:54:18 | crate | None | None |
| canonical_paths.rs:54:14:54:35 | crate::canonical_paths | repo::test | crate::canonical_paths |
| canonical_paths.rs:54:14:54:38 | crate::canonical_paths::a | repo::test | crate::canonical_paths::a |
| canonical_paths.rs:54:14:54:45 | crate::canonical_paths::a::Trait | repo::test | crate::canonical_paths::a::Trait |
| canonical_paths.rs:54:51:54:61 | OtherStruct | repo::test | {34}::OtherStruct |
| canonical_paths.rs:63:21:63:31 | OtherStruct | repo::test | {34}::OtherStruct |
| canonical_paths.rs:64:13:64:13 | s | None | None |
| canonical_paths.rs:64:13:64:17 | ... .f(...) | repo::test | <{34}::OtherStruct as crate::canonical_paths::a::Trait>::f |
| canonical_paths.rs:65:13:65:13 | s | None | None |
| canonical_paths.rs:65:13:65:17 | ... .g(...) | repo::test | <{34}::OtherStruct as {34}::OtherTrait>::g |
| canonical_paths.rs:66:13:66:18 | nested | repo::test | {34}::nested |
